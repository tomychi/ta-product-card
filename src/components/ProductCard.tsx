import React, { createContext } from 'react';
import { useProduct } from '../hooks/useProducts';
import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';


import styles from '../styles/styles.module.css';

// creamos nuestro contexto y definimos como va a ser
export const ProductContext = createContext({} as ProductContextProps);
// extraemos el provider; proveedor de informacion
const { Provider } = ProductContext;

// por si recibo mas de una propiedad
export interface Props {
    product: Product;
    // children?: React.ReactElement | React.ReactElement[]; // si envio varios hijos
    children: ( args: ProductCardHandlers )=> JSX.Element;
    className?: string;
    style?: React.CSSProperties;
    onChange?: ( args: onChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues;
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {

    // para que todos los cambios los maneje useProduct le mando el onChange como dependencia
    const { counter, increaseBy, maxQuantity, isMaxQuantityReached, reset } = useProduct( { onChange, product, value, initialValues  } );

    return (
        <Provider
            value={{
                product,
                counter,
                maxQuantity,
                increaseBy,
            }}
        >
            <div 
                className={`${styles.productCard} ${className}`}
                style={style}
            >
                { 
                    children({
                        quantity: counter,
                        isMaxQuantityReached,
                        maxQuantity: initialValues?.maxQuantity,
                        product,

                        increaseBy,
                        reset,
                    })
                }
            </div>
        </Provider>
    );
};
