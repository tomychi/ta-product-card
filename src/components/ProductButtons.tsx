import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';
import { useCallback } from "react";
export interface Props {
    className?: string;
    style?: React.CSSProperties;
}

export const  ProductButtons = ({className, style }: Props) => {

    // extraemos la informacion del contexto
    const {increaseBy, counter, maxQuantity  } = useContext(ProductContext);

    // TODO: isMaxReached = useCallback( () => {}, [maxQuantity, counter] );
    // TRUE si el count === maxQuantity
    // FALSE si no lo es
    const isMaxReached = useCallback(
        // si no existe 
        () => !!maxQuantity && counter === maxQuantity,
        [counter, maxQuantity],
    )
    


    return (
        <div 
            className={ `${styles.buttonsContainer} ${className}` }
            style={style}
        >

            <button 
                className={ styles.buttonMinus } 
                onClick={ () => increaseBy(-1) }
            > 
                - 
            </button>

            <div 
                className={ styles.countLabel }
            > 
                { counter } 
            </div>

            <button 
                className={ `${styles.buttonAdd} ${ isMaxReached() && styles.disabled}`}
                onClick={ () => increaseBy(+1) }
            > 
                + 
            </button>

        </div>
    );

}
