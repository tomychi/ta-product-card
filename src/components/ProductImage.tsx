import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";


import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';

export interface Props {
    className?: string;
    img?: string;
    style?: React.CSSProperties;
}


export const ProductImage = ({img = '', className, style}: Props ) => {

    const { product } = useContext(ProductContext);
    let imgToShow: string;

    if (img) {
        imgToShow = img;
    } else if (product.img) {
        imgToShow = product.img;
    } else {
        imgToShow = noImage;
    }

    return (
        <img  
            className={ `${styles.productImg} ${className}` }
            style={style}
            // si viene la imagen la muestro, sino la imagen por defecto
            src={ imgToShow } 
            alt="Product images" 
        />
    )
}