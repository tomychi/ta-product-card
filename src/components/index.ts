import { ProductCard as ProductCardHOC } from './ProductCard';

import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductCardHOCProps } from '../interfaces/interfaces';

export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';

// forma de propiedades
export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Buttons: ProductButtons,
    Image: ProductImage,
    Title: ProductTitle,
});

export default ProductCard;
