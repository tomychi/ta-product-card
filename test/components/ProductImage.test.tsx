import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('debe mostrar el componente correctamente con la iamgen', () => {
    const wrapper = renderer.create(<ProductImage img="https://hola.jpg" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe de mostar el componente con la iamgen del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
