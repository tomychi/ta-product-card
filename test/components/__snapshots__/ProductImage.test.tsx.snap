// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe de mostar el componente con la iamgen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product images"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage debe mostrar el componente correctamente con la iamgen 1`] = `
<img
  alt="Product images"
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;
